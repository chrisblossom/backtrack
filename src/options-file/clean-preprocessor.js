/* @flow */

import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import type { Clean, NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value?: Clean | $ReadOnlyArray<Clean>,
};

function cleanPreprocessor({ value }: Args = {}): $ReadOnlyArray<
    NormalizedClean,
> {
    const allToArray = toArray(value);

    return allToArray.map((arg: Clean) => {
        const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
            const resolvedMakeDir = path.resolve(buildPath, makeDir);
            return resolvedMakeDir;
        });

        const copy = toArray(arg.copy).map((copyArgs) => {
            const resolvedSrc = path.resolve(rootPath, copyArgs.src);
            const resolvedDest = path.resolve(buildPath, copyArgs.dest);

            const result = {
                src: resolvedSrc,
                dest: resolvedDest,
                hash: copyArgs.hash,
            };

            if (result.hash === undefined) {
                delete result.hash;
            }

            return result;
        });

        return {
            del: toArray(arg.del),
            makeDirs,
            copy,
        };
    });
}

export { cleanPreprocessor };
