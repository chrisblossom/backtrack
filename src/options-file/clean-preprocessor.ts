import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import { Clean, NormalizedClean } from '../types';

interface Args {
	value?: Clean | Clean[];
}

function cleanPreprocessor({ value }: Args = {}): NormalizedClean[] {
	const allToArray = toArray(value);

	return allToArray.map((arg: Clean) => {
		const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
			const resolvedMakeDir = path.resolve(buildPath, makeDir);
			return resolvedMakeDir;
		});

		const copy = toArray(arg.copy).map((copyArgs) => {
			const resolvedSrc = path.resolve(rootPath, copyArgs.src);
			const resolvedDest = path.resolve(buildPath, copyArgs.dest);

			const result = {
				src: resolvedSrc,
				dest: resolvedDest,
				hash: copyArgs.hash,
			};

			if (result.hash === undefined) {
				delete result.hash;
			}

			return result;
		});

		return {
			del: toArray(arg.del),
			makeDirs,
			copy,
		};
	});
}

export { cleanPreprocessor };
