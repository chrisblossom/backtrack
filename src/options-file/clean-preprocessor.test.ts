const cleanPreprocessor = (args?: any) =>
	require('./clean-preprocessor').cleanPreprocessor(args);

describe('cleanPreprocessor', () => {
	test('handles undefined', () => {
		const result = cleanPreprocessor();

		expect(result).toEqual([]);
	});

	test('converts all values to arrays and adds missing keys', () => {
		const value = [
			{
				del: '*',
			},
			{
				makeDirs: 'test',
			},
			{
				copy: {
					src: 'static-1',
					dest: 'static-1',
					hash: true,
				},
			},
			{
				copy: {
					src: 'static-2',
					dest: 'static-2',
				},
			},
			{},
			{
				del: '*',
				makeDirs: 'test',
				copy: {
					src: 'static-3',
					dest: 'static-3',
				},
			},
		];

		const result = cleanPreprocessor({ value });

		expect(result).toMatchSnapshot();
	});
});

// Use an empty export to please Babel's single file emit.
// https://github.com/Microsoft/TypeScript/issues/15230
export {};
