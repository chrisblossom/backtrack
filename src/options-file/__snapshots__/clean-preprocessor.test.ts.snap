// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanPreprocessor converts all values to arrays and adds missing keys 1`] = `
[
  {
    "copy": [],
    "del": [
      "*",
    ],
    "makeDirs": [],
  },
  {
    "copy": [],
    "del": [],
    "makeDirs": [
      "<PROJECT_ROOT>/dist/test",
    ],
  },
  {
    "copy": [
      {
        "dest": "<PROJECT_ROOT>/dist/static-1",
        "hash": true,
        "src": "<PROJECT_ROOT>/static-1",
      },
    ],
    "del": [],
    "makeDirs": [],
  },
  {
    "copy": [
      {
        "dest": "<PROJECT_ROOT>/dist/static-2",
        "src": "<PROJECT_ROOT>/static-2",
      },
    ],
    "del": [],
    "makeDirs": [],
  },
  {
    "copy": [],
    "del": [],
    "makeDirs": [],
  },
  {
    "copy": [
      {
        "dest": "<PROJECT_ROOT>/dist/static-3",
        "src": "<PROJECT_ROOT>/static-3",
      },
    ],
    "del": [
      "*",
    ],
    "makeDirs": [
      "<PROJECT_ROOT>/dist/test",
    ],
  },
]
`;
